integrations/opennms-geoip-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/GeoIpProvisioningAdapter.java

```diff
@@ -0,0 +1,327 @@
+import com.atomikos.beans.PropertyUtils;
+import com.google.common.base.Strings;
+import com.maxmind.geoip2.DatabaseReader;
+import com.maxmind.geoip2.exception.GeoIp2Exception;
+import com.maxmind.geoip2.model.CityResponse;
+
+@EventListener(name = GeoIpProvisioningAdapter.NAME)
+public class GeoIpProvisioningAdapter extends SimplerQueuedProvisioningAdapter implements InitializingBean {
+    private static final Logger LOG = LoggerFactory.getLogger(GeoIpProvisioningAdapter.class);
+    public static final String PREFIX = "Provisiond.";
+    public static final String NAME = "GeoIpProvisioningAdapter";
+
+    private NodeDao nodeDao;
+    private EventForwarder eventForwarder;
+    private GeoIpConfigDao geoIpConfigDao;
+    private DatabaseReader databaseReader;
+    private AssetRecordDao assetRecordDao;
+
+    public GeoIpProvisioningAdapter() {
+        super(NAME);
+    }
+
+    private DatabaseReader getDatabaseReader() throws IOException {
+        if (databaseReader == null) {
+            final GeoIpConfig geoIpConfig = geoIpConfigDao.getContainer().getObject();
+            databaseReader = new DatabaseReader.Builder(new File(geoIpConfig.getDatabase())).build();
+        }
+
+        return databaseReader;
+    }
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        Assert.notNull(nodeDao, "Node DAO must not be null");
+        Assert.notNull(assetRecordDao, "Asset Record DAO must not be null");
+        Assert.notNull(eventForwarder, "Event Forwarder must not be null");
+        Assert.notNull(geoIpConfigDao, "GeoIp Configuration DAO must not be null");
+    }
+
+    @Override
+    public void doAddNode(final int nodeId) throws ProvisioningAdapterException {
+        queryNode(nodeId);
+    }
+
+    @Override
+    public void doUpdateNode(final int nodeId) throws ProvisioningAdapterException {
+        queryNode(nodeId);
+    }
+
+    private boolean isPublicAddress(final InetAddress address) {
+        return !(address.isSiteLocalAddress() ||
+                address.isAnyLocalAddress() ||
+                address.isLinkLocalAddress() ||
+                address.isLoopbackAddress() ||
+                address.isMulticastAddress());
+    }
+
+    public void queryNode(final int nodeId) {
+        final GeoIpConfig geoIpConfig = geoIpConfigDao.getContainer().getObject();
+
+        if (!geoIpConfig.isEnabled()) {
+            return;
+        }
+
+        final OnmsNode node = nodeDao.get(nodeId);
+        if (node == null) {
+            LOG.debug("Failed to return node for given nodeId: {}", nodeId);
+            return;
+        }
+
+        OnmsAssetRecord assetRecord = node.getAssetRecord();
+
+        if (assetRecord == null) {
+            assetRecord = new OnmsAssetRecord();
+            assetRecord.setNode(node);
+        }
+
+        InetAddress addressToUse = node.getPrimaryInterface().getIpAddress();
+
+        if (geoIpConfig.getResolve().startsWith("public")) {
```

Comment:
Print out a warning for an invalid value from `getResolve()`. Can this be an `enum`?

Reply:
done